Track `-Cprofile-use` and `-Cprofile-sample-use` value by file hash, not file path

Kobzol opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/100413)*
<!-- homu-ignore:end -->

Before, the path to PGO profile passed in `-Cprofile-use` and `-Cprofile-sample-use` was tracked by the filepath only. This meant that if the code was compiled twice in a row with the same path, the crate would not be recompiled, even if the profile content has changed in the meantime.

I'm not too excited about the used `md-5` crate's API, but it was already used for hashing source files, so I decided to keep the same dependency. I'm not really sure what the `for_crate_hash` argument is for, should I take it into account here?

Fixes: https://github.com/rust-lang/rust/issues/100397

r? @michaelwoerister

Kobzol committed:
Track `-Cprofile-use` and `-Cprofile-sample-use` values by file hash, not file path

bors commented:
:umbrella: The latest upstream changes (presumably #100595) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7853181512?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/db05f78310e47406522fa17ee19803fc5bb3e326/checks/7853181512/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Finished dev [unoptimized] target(s) in 0.04s
Skipping Set({test::src/tools/tidy}) because it is excluded
Skipping Set({test::src/tools/tidy}) because it is excluded
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --locked flag and use --offline instead.
```

</details>

michaelwoerister reviewed (commented):
Whoops, I just discovered that my comments from a few days ago are still "pending". So, here goes:

michaelwoerister commented:
Overall, I'm not sure about the approach. Maybe `HashedFilePath` could cache the fingerprint upon construction? 

Another way to do it would be to turn retrieving the paths into `eval_always` queries. E.g.
```
query pgo_use_path(_: ()) -> ContentHashedFilePath {
    eval_always
    desc { "path to .profdata file used by LLVM" }
}

#[derive(HashStable)]
struct ContentHashedFilePath {
    path: PathBuf,
    /// A fingerprint of the file contents that `path` points to.
    hash: Fingerprint,
}
```

The `HashedFilePath` struct used in Session could then mark accesses to the path as `unsafe`, so that it's unlikely that someone accidentally uses it instead of the query.

The advantage of that approach is that the query system will take care of caching the result of the expensive operation, and that changes to the contents of the `.profdata` file would not invalidate the whole session, just the codegen part. (Not sure how likely that is in practice though)

The downside is that it's quite a bit more complex.

Kobzol commented on compiler/rustc_session/src/utils.rs:
```diff
@@ -91,3 +100,67 @@ impl CanonicalizedPath {
+}
+
+fn hash_file(path: &Path) -> std::io::Result<Output<Md5>> {
+    let mut hasher = Md5::default();
```
Hmm, AFAIK `StableHasher` is optimized for frequent and small updates, while here we want to hash large chunks of bytes.

I originally wanted to use Blake3, but decided on MD5 to avoid adding a new dependency. I created a small benchmark, which hashed a 4GiB profile file with random data (primed in file cache, so hopefully without much I/O overhead) with 128 KiB block size:
```
MD5: 9200 - 10000ms
BLAKE3: 1099 - 1100ms
StableHasher: 1687 - 1744ms
```

BLAKE3 looks like a good choice, especially since it was designed for use-cases like hashing files. But the stable hasher also isn't half bad.

Kobzol replied:
(Btw I also tried swapping the hashing algorithm for source files for blake3, but it didn't really help. Well, most source files don't have 4 GiB :) ).

michaelwoerister replied:
XXH3 would be the ideal choice ;)

Do we already have a dependency on BLAKE3? I agree that it looks like a great choice!

Kobzol replied:
We don't, I would have to add it to `rustc_session`. It uses `CC0 1.0`/`Apache 2.0` license.

Kobzol replied:
Btw, when we do PGO for LLVM in CI, all the files have 16 GiB, but after they are combined into a single file, it's just under 30 MiB. For `rustc` it's 80 MiB. So maybe large optimizations here are unnecessary and we can just use `StableHasher` to avoid a new dependency.

michaelwoerister replied:
> all the files have 16 GiB
😅 

I'd just go with StableHasher for now, I think. We can always switch later.

Kobzol commented on compiler/rustc_session/src/utils.rs:
```diff
@@ -91,3 +100,67 @@ impl CanonicalizedPath {
+    let mut hasher = Md5::default();
+
+    let mut file = File::open(path)?;
+    let mut buffer = [0; 4096];
```
Higher block probably means worse L1 cache reuse, but 128 KiB should be fine. I'll change it to this size.

Kobzol commented:
> The advantage of that approach is that the query system will take care of caching the result of the expensive operation, and that changes to the contents of the .profdata file would not invalidate the whole session, just the codegen part. (Not sure how likely that is in practice though)

This could actually be quite useful, because now if the profile changes, it will recompile all dependencies, which can be quite time consuming. Would this invalidation also works for depedencies? In other words, if I implement this query and the profile changes, will it only do codegen for dependencies, without recompiling the rest of the dependencies?

michaelwoerister commented:
> In other words, if I implement this query and the profile changes, will it only do codegen for dependencies, without recompiling the rest of the dependencies?

IIRC, Cargo does not compile dependencies incrementally because it makes the assumption that those change very infrequently. Under that assumption, the query approach would not stop dependencies from being recompiled completely. I'm not sure if the CARGO_INCREMENTAL env var has any influence on this.

You could check by running Cargo with `--verbose` and see if `-Cincremental` shows up in the rustc commandline for dependencies.

Kobzol commented:
> You could check by running Cargo with --verbose and see if -Cincremental shows up in the rustc commandline for dependencies.

Indeed `-C incremental` is not there, so this benefit wouldn't work. Then I wonder if we should just use the simpler approach and use something like `Option<Hash>` to cache the hash inside the hashed file struct?

michaelwoerister commented:
> Then I wonder if we should just use the simpler approach and use something like Option<Hash> to cache the hash inside the hashed file struct?

Yes, let's do that. `HashedFilePath::new()` can do the hashing, right? 

Maybe I'd rename `HashedFilePath` to `ContentHashedFilePath`, btw, just to make it clearer what's going on.

Kobzol commented:
> Yes, let's do that. HashedFilePath::new() can do the hashing, right?

Indeed, that's a good idea. Before if the file changed during the compilation, it could be hashed several times, each time with a different hash, which is bad.

I renamed the struct, switched to `StableHasher` and changed the code to compute the hash in `ContentHashedFilePath::new()`.

michaelwoerister commented:
I just realized that the current implementation will cause the file to be hashed in non-incremental mode too :/ That doesn't seem optimal, right?

Kobzol commented:
Since the `LLVM`/`rustc` profiles from our CI, which are quite large in their raw form, result in a 30 - 80 MiB file after merge, and only the merged file then goes to PGO (the raw files always have to be merged first AFAIK), I'd say that it's not such a big deal and in most cases the hashing will be very fast.

But even then, I think that we should hash the profile even with non-incremental mode. Consider this:
```bash
$ CARGO_INCREMENTAL=0 RUSTFLAGS="-Cprofile-use=profile.prof" cargo build --release
# Change profile.prof
$ CARGO_INCREMENTAL=0 RUSTFLAGS="-Cprofile-use=profile.prof" cargo build --release
# In current rustc, the crate is not recompiled!
```

I think that we should invalidate the profile even in non-incremental mode, or not?

michaelwoerister commented:
Decisions about recompiling something or not are done by Cargo. The hash we are talking about does not influence anything in non-incremental mode, if I understand correctly. Cargo would need to recognize the `.profdata` file as an input to `rustc`. I'm not sure if it tries to parse out information from `RUSTFLAGS`.

michaelwoerister commented:
To elaborate, there's a two-level process involved here:
1. Cargo looks at the filesystem timestamps of compilation inputs to decide whether to invoke rustc or not.
2. rustc might use the incr. comp. cache to do less work, if it is invoked incrementally.

Kobzol commented:
Aha. Well, but this fact makes this whole PR obsolete :) Since Cargo won't even invoke `rustc`, the code will not get recompiled even if the profile changes. I tried it now with this PR and indeed Cargo doesn't even invoke `rustc`, so the code doesn't get recompiled when the profile file changes.

michaelwoerister commented:
I don't think the PR is obsolete -- we still need to deal with the cases where rustc does get invoked with `-Cincremental` so we don't re-use out-dated object files. 
But yes, Cargo also needs to be made aware that the `.profdata` is an input to compilation. The Cargo folks would know more about how to do that cleanly.

Kobzol commented:
I discussed this on the Cargo zulip stream (https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Tracking.20PGO.20profile.20files.20in.20cargo) and it seems that there's a way forward. Since this change is relatively self-contained, should we merge it now? Or do you want me to make the Cargo related changes in this PR too?

michaelwoerister commented:
I think the cargo related changes should go into a separate PR.

Regarding this PR: I'm still not really happy about hashing the file in non-incremental mode. I'll think about it some more over the weekend.

bors commented:
:umbrella: The latest upstream changes (presumably #101577) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

michaelwoerister commented:
@Kobzol, if you are still interested in this: I think the best way forward is to handle this in the query system. That way we won't do any unnecessary work in non-incremental and in incremental mode, invalidation will be limited to just the LLVM part. 

I imagine it to work like this:
- Add queries for the profile-use and profile-sample-use paths. The query provider reads the path from the session (might be good to somehow poison the field in the session at that point, so it's not accessible anymore -- but this is a more general problem).
- The result type of these queries holds the path but its `HashStable` implementation will hash the contents of the file (we have to be careful to not hash the file multiple times here).
- we add a field for the PGO paths to the [ModuleCodegen struct](https://cs.github.com/rust-lang/rust/blob/58546803885164d488185fb9cb9fb04fcbe64e30/compiler/rustc_codegen_ssa/src/lib.rs#L53), so backends have to access the queries, which will add the appropriate dependency edges.

The rest would be handled by the existing incr. comp. infrastructure. 
What do you think?

apiraino commented:
Based on [this comment](https://github.com/rust-lang/rust/pull/100413#issuecomment-1278104829) I'll tentatively switch review status. Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

Dylan-DPC commented:
@kobzol any updates on this?

michaelwoerister commented:
I'll be looking at an issue with `#[debugger_visualizer]` that might be somewhat similar in nature. Maybe fixing that might introduce some infrastructure that could also be used here.

I still think that the issue this PR addresses should be fixed eventually and I don't mind leaving it open and assigned to me as a reminder. Unless of course @Kobzol wants to get rid of the open PR 🙂 Then we can migrate this to a bug report.

Dylan-DPC commented:
Hmm. In that case i'll mark it as blocked till that work is done

michaelwoerister commented:
I'm unassigning myself (see https://github.com/rust-lang/team/pull/1565). I think the general outline in https://github.com/rust-lang/rust/pull/100413#issuecomment-1271491267 would still be a valid approach to solve the problem.

jieyouxu commented:
r? incremental (in case whoever is more familiar with incremental have any advice)
